src/base_alloc/base_alloc.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+
+#include "base_alloc.h"
+#include "base_alloc_internal.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+
+// minimum size of a single pool of the base allocator,
+#define MINIMUM_POOL_SIZE (ba_os_get_page_size())
+
+// minimum number of chunks used to calculate the size of pools
+#define MINIMUM_CHUNK_COUNT (128)
+
+// alignment of the base allocator
+#define MEMORY_ALIGNMENT (8)
+
+typedef struct ba_mem_chunk_t ba_mem_chunk_t;
+
+// memory chunk of size 'chunk_size'
+struct ba_mem_chunk_t {
+    ba_mem_chunk_t *next;
+};
+
+struct umf_ba_pool_meta_t {
+    // address of the beginning of this pool (allocated region)
+    umf_ba_pool_t *pool_addr;
+    // address of the beginning of the next pool (a list of allocated pools to be freed in umf_ba_destroy())
+    umf_ba_pool_t *next_pool;
+
+    // the following fields are set and used only in the main (the first) pool
+    size_t pool_size; // size of each pool (argument of each ba_os_alloc() call)
+    size_t pool_usable_size;   // usable size of each pool
```

Comment:
I think pool_usable_size is not needed at all in this struct. It doesn't change after allocating new pool, right? It could be calculated when needed from other constants.

Reply:
Done